### PR TITLE
Fix ANGLE/OpenGL swap buffer issues under Windows

### PR DESCRIPTION
**Pull request by JoergAtGithub**
This fixes the bugs https://bugs.launchpad.net/mixxx/+bug/1895427 and https://bugs.launchpad.net/mixxx/+bug/1872172 under the OpenGL emulation ANGLE.

I tested this fix with ANGLE the angle library, that comes with 2.3-j00019-x64-release-fastbuild-static-55e94982-minimal as part of the Qt 5.14.2.
It also works if I set QT_OPENGL to desktop to use the native OpenGL driver of my Intel HD Graphics 3000 GPU.

I can't test the behaviour on other platforms.

Furthermore it fixes issue https://bugs.launchpad.net/mixxx/+bug/1905123 by adding the missing ANGLE DLLs to the windows installer.

### CODE REVIEW

**Comment by Be-ing** [comment]
Yay! I'm glad you found a solution that doesn't require any hacks overriding the ANGLE version included in Qt.

**Comment by Be-ing** [comment]
We still need to copy the ANGLE DLLs into the installer right? Can you add `install` commands to CMakeLists.txt to do that?

File: CMakeLists.txt [hunk 1]
```diff
@@ -1148,6 +1148,15 @@ install(
...
 )
+
+# Qt loads the ANGLE libraries dynamical, also when Qt itself is linked static 
+install(
+  FILES
+    "${Qt5_DIR}/bin/libEGL.dll"
+    "${Qt5_DIR}/bin/libGLESv2.dll"
+  DESTINATION
+    "${MIXXX_INSTALL_DATADIR}"
+)
```
Comment (Be-ing) [inline CMakeLists.txt:1159]: Put this within the `if(WIN32)` block above.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1148,6 +1148,15 @@ install(
   FILES_MATCHING REGEX
     "qt_.+\.qm|qtbase_.*\.qm|qtmultimedia_.*\.qm|qtscript_.*\.qm|qtxmlpatterns_.*\.qm"
 )
+
+# Qt loads the ANGLE libraries dynamical, also when Qt itself is linked static
```
Comment (Be-ing) [inline CMakeLists.txt:1152]:
```suggestion
# Qt loads the ANGLE libraries dynamically, even when Qt itself is linked statically 
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -1148,6 +1148,15 @@ install(
...
 )
+
+# Qt loads the ANGLE libraries dynamically, even when Qt itself is linked statically
+install(
+  FILES
+    "${Qt5_DIR}/bin/libEGL.dll"
+    "${Qt5_DIR}/bin/libGLESv2.dll"
+  DESTINATION
+    "${MIXXX_INSTALL_DATADIR}"
+)
```
Comment (Be-ing) [inline CMakeLists.txt:1159]: Please fix the indentation here. The `install` block above is also not indented properly.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1135,28 +1135,28 @@ install(
...
 
-# Qt loads the ANGLE libraries dynamically, even when Qt itself is linked statically
-install(
-  FILES
-    "${Qt5_DIR}/bin/libEGL.dll"
-    "${Qt5_DIR}/bin/libGLESv2.dll"
-  DESTINATION
-    "${MIXXX_INSTALL_DATADIR}"
-)
+    # Qt loads the ANGLE libraries dynamically, even when Qt itself is linked statically
```
Comment (Be-ing) [inline CMakeLists.txt:1152]: The rest of CMakeLists.txt uses 2 spaces for each indentation level. Please do the same here.

File: src/waveform/waveformwidgetfactory.cpp [hunk 5]
```diff
@@ -713,6 +713,7 @@ void WaveformWidgetFactory::swap() {
                 }
                 QGLWidget* glw = qobject_cast<QGLWidget*>(pWaveformWidget->getWidget());
                 if (glw != nullptr) {
+                    glw->makeCurrent();
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp:716]: Does it still work when guarding it with: 
if (QGLContext::currentContext() != context()) {

**Review by daschuer — COMMENTED** [review]
Thank you for do all the hard debugging and discover this. 
I will test this on Linux tomorrow.

**Comment by Be-ing** [comment]
Please merge or rebase onto 2.3 to include #3368. I merged these two in my own branch to try to get GitHub Actions to build an artifact to test. However, the new `install` command in CMakeLists.txt is breaking cpack:
https://github.com/Be-ing/mixxx/runs/1450285227
```
CPack: Create package using WIX
CPack: Install projects
CPack: - Install project: mixxx
CMake Error at D:/a/mixxx/mixxx/cmake_build/cmake_install.cmake:56 (file):
  file INSTALL cannot find "/bin/libEGL.dll".


CPack Error: Error when generating package: mixxx
```

**Comment by uklotzde** [comment]
Rebase and stash the commits.

**Comment by Be-ing** [comment]
I searched CMakeLists.txt for `egl` and found this interesting bit:
```
    target_link_libraries(mixxx-lib PUBLIC
      # Pulled from qt-4.8.2-source\mkspecs\win32-msvc2010\qmake.conf
      # QtCore
      kernel32
      user32      # QtGui, QtOpenGL, libHSS1394
      shell32
      uuid
      ole32       # QtGui,
      advapi32    # QtGui, portaudio, portmidi
      ws2_32      # QtGui, QtNetwork, libshout
      # QtGui
      gdi32       # QtOpenGL, libshout
      comdlg32
      oleaut32
      imm32
      winmm
      winspool
      # QtOpenGL
      glu32
      opengl32

      # QtNetwork openssl-linked
      crypt32

      # New libraries required by Qt5.
      dwmapi      # qtwindows
      iphlpapi    # qt5network
      #libEGL      # qt5opengl
      #libGLESv2   # qt5opengl
      mpr         # qt5core
      netapi32    # qt5core
      userenv     # qt5core
      uxtheme     # ?
      version     # ?
      wtsapi32    # ?

      # NOTE(rryan): If you are adding a plugin here, you must also
      # update src/mixxxapplication.cpp to define a Q_IMPORT_PLUGIN
      # for it. Not all imageformats plugins are built as .libs when
      # building Qt statically on Windows. Check the build environment
      # to see exactly what's available as a standalone .lib vs linked
      # into Qt .libs by default.

      # iconengines plugins
      Qt5::QSvgIconPlugin

      # imageformats plugins
      Qt5::QGifPlugin
      Qt5::QICOPlugin
      Qt5::QJpegPlugin
      Qt5::QSvgPlugin
      Qt5::QTgaPlugin

      # platform plugins (new in Qt5 for Windows)
      Qt5::QWindowsIntegrationPlugin

      # styles (new in Qt5 for Windows)
      Qt5::QWindowsVistaStylePlugin

      # sqldrivers (new in Qt5? or did we just start enabling them)
      Qt5::QSQLiteDriverPlugin
    )
```

Should we be using `target_link_libraries` to dynamically link libEGL and libGLESv2 instead of manually `install`ing the DLLs? @Holzhaus what do you think?

File: src/waveform/waveformwidgetfactory.cpp (hunk 5 above)
Comment (JoergAtGithub) [reply src/waveform/waveformwidgetfactory.cpp:716]: What should be context() in this statement? I didn't get the context here;-)

Comment (daschuer) [reply src/waveform/waveformwidgetfactory.cpp:716]: ups, glw->context()

Comment (JoergAtGithub) [reply src/waveform/waveformwidgetfactory.cpp:716]: This code seems also work without failures: grafik

Comment (Be-ing) [reply src/waveform/waveformwidgetfactory.cpp:716]: Combine it with the `if` condition above: `if (glw != nullptr && glw->context() != QGLContext::curentContext()) {`

Comment (Be-ing) [reply src/waveform/waveformwidgetfactory.cpp:716]: nevermind, that isn't logically equivalent

Comment (Be-ing) [reply src/waveform/waveformwidgetfactory.cpp:716]: @daschuer does `glw->makeCurrent()` have any performance penalty if the context is already current? Is the check for `glw->context() != QGLContext::currentContext()` needed?

Comment (JoergAtGithub) [reply src/waveform/waveformwidgetfactory.cpp:716]: The problem seems to be, that the openGL context is sometimes not set to the correct thread: https://doc.qt.io/qt-5/qopenglcontext.html#makeCurrent
In this cases swapBuffers fails, but swapBuffers has to be executed always.

**Comment by Holzhaus** [comment]
> Should we be using `target_link_libraries` to dynamically link libEGL and libGLESv2 instead of manually `install`ing the DLLs? @Holzhaus what do you think?

I honestly have no idea. Are they always present on windows (i.e. system libraries)?

**Comment by Be-ing** [comment]
No, libEGL.dll and libGLESv2 come from ANGLE which is included in Qt. They are hacks around the OS not properly supporting OpenGL.

**Comment by Holzhaus** [comment]
If we link them dynamically without shipping them, won't it cause issues because the windows loader will fail to locate the DLLs? I have no idea how all that stuff works on windows. If you think it's an improvement, try it.

**Comment by Be-ing** [comment]
cpack failed again:
```
Run cpack -G WIX
CPack: Create package using WIX
CPack: Install projects
CPack: - Install project: mixxx
CMake Error at D:/a/mixxx/mixxx/cmake_build/cmake_install.cmake:56 (file):
  file INSTALL cannot find "/../../../bin/libEGL.dll".
```
It seems Qt5_DIR is not initialized at the point where the `install` call is made. I think it is initialized by `find_package(Qt5` lower down in CMakeLists.txt.

I think it is worth a try uncommenting `libEGL` and `libGLESv2` from the `target_link_libraries(mixxx-lib PUBLIC` call. It might be necessary to get rid of the `lib` prefix.

**Comment by Be-ing** [comment]
> If we link them dynamically without shipping them

I am thinking cpack might automatically include them in the MSI installer if they are linked by cmake. I could be wrong about that though.

**Comment by Holzhaus** [comment]
:shrug:

File: src/waveform/waveformwidgetfactory.cpp (hunk 5 above)
Comment (daschuer) [reply src/waveform/waveformwidgetfactory.cpp:716]: > @daschuer does glw->makeCurrent() have any performance penalty if the context is already current? Is the check for glw->context() != QGLContext::currentContext() needed?

makeCurrent() is expensive. We have the check everywhere else for some reasons. 

Maybe this is done internally anyway. in this case we can omit the check. 
I need to debug into the code to be sure.  

The other puzzling question is why it works with real hardware. Maybe we are doing something wrong that steals the context from the thread in the Angle case.

Comment (Be-ing) [reply src/waveform/waveformwidgetfactory.cpp:716]: > Maybe this is done internally anyway. 

I checked. It is not. We should do the check before calling makeCurrent.
https://code.woboq.org/qt5/qtbase/src/opengl/qgl.cpp.html#_ZN10QGLContext11makeCurrentEv

**Comment by Be-ing** [comment]
> Should we be using target_link_libraries to dynamically link libEGL and libGLESv2 instead of manually installing the DLLs?

I tested this and it does not work. Linking fails.

**Comment by Be-ing** [comment]
I fixed the CMakeLists.txt and cleaned up the Git history in #3370.

**Comment by Be-ing** [comment]
Thank you @JoergAtGithub for working on this!

**Pull request closed without merging**